Zend/zend_list.c

```diff
@@ -216,10 +216,11 @@ void zend_close_rsrc_list(HashTable *ht)
 {
 	/* Reload ht->arData on each iteration, as it may be reallocated. */
 	uint32_t i = ht->nNumUsed;
+
 	while (i-- > 0) {
-		Bucket *p = &ht->arData[i];
-		if (Z_TYPE(p->val) != IS_UNDEF) {
-			zend_resource *res = Z_PTR(p->val);
+		zval *p = HT_IS_PACKED(ht) ? &ht->arPacked[i] : &ht->arData[i].val;
```

Comment:
`/* Reload ht->arData on each iteration, as it may be reallocated. */`

I assume the only possible allocation is to grow the array, because resource ids can't be reused? Would be useful to document that implicit assumption if that was the case (to prevent people from copying this to places where that's wrong and size can shrink (and `i` would be out of bounds)).

Would code like this that iterates over values in raw buckets may benefit from a macro when it isn't performance sensitive enough to use two different loops (also to reduce maintenance work), e.g. in xdebug, phpdbg, uopz/runkit7, serializers, etc.?

```php
#define HT_GET_ZVAL_AT_RAW_OFFSET(ht, i) (HT_IS_PACKED((ht)) ? &(ht)->arPacked[i] : &(ht)->arData[i].val)
```

```suggestion
		zval *p = HT_GET_ZVAL_AT_RAW_OFFSET(i);
```

Reply:
@ TysonAndre In general, packed array may be converted into real hash with less number of elements.